### DISCUSSION TITLE
Add info (metadata) with JSON to PDF without any yet

### DISCUSSION THREAD

**Discussion by SeeLook** [qpdf/qpdf#1309] [Q&A] [answered]
Hi all,

How to add metadata (title, author, etc.) using `--update-from-json`  to PDF which doesn't contain any such an object yet.  
I'm able to do that if the input PDF contains already an object with any metadata tag. So adding:
```json
{
    "qpdf": [
        {
            "jsonversion": 2
        },
        {
            "obj:2 0 R": {
                "value": {
                    "/CreationDate": "u:D:20241114231320+00'00'",
                    "/ModDate": "u:D:20241114231320+00'00'",
                    "/Producer": "u:Some App"
                }
            }
        }
    ]
}

```
does the job.  
How to do this when PDF doesn't contain such a data so there is no object to replace.

**Comment by jberkenbilt** [comment]
You can just add a new object with the next available object number and hook it into the /Root object by adding an /Info key that points to it. In the qpdf json output, in the first element, there is `maxobjectid`, which is the highest existing object ID. So if that's 30, you can create `obj:31 0 R` with your Info dictionary and add `"/Info": "31 0 R"` to your root. Let me know if that's not enough information, and I can spell it out a little more explicitly. I feel like I documented this somewhere, but I wasn't able to find it quickly in the docs or in my blog.

**Reply by SeeLook** [reply]
Thank You a lot for the answer.  
Let me rephrase it. I need to squeeze  this `"/Info": "Obj:31 0 R"` into trailer:  
```json
"trailer": {
        "value": {
          "/ID": [
            "b:06528ffe7b536973fba69940af3c6714",
            "b:06528ffe7b536973fba69940af3c6714"
          ],
          "/Root": "1 0 R",
          "/Size": 30
        }
      }
```
This above example is exactly trailer from the PDF without metadata.  
Is it right?

**Reply by SeeLook** [reply]
@jberkenbilt please give more specific instructions.  
I was trying to put `"/Info": "Obj:31 0 R"` key in many places but nothing works.  
Also I can't find any clue in a doc.

**Reply by jberkenbilt** [reply]
`/Info` is a field of the root dictionary, not the trailer, so you'll want to add the field to "obj:1 0 R". You can take a look at another PDF as an example. Feel free to ask any additional questions.

**Reply by jberkenbilt** [reply]
It will be "/Info": "31 0 R". The `obj:...` form is only used in keys.

**Reply by SeeLook** [reply, marked as answer]
This combination works for me:
```json
{
    "qpdf": [
        {
            "jsonversion": 2
        },
        {
            "obj:30 0 R": {
                "value": {
                    "/CreationDate": "u:D:20241114231320+00'00'",
                    "/ModDate": "u:D:20241114231320+00'00'",
                    "/Producer": "u:Some App"
                }
            },
            "trailer": {
              "value": {
                  "/Info": "30 0 R",
                  "/Root": "1 0 R"
              }
            }
        }
    ]
}
```
This `/Info` with `obj` I copied by mistake. Sorry for mess up.
And first of all thank You a lot for help.

**Reply by jberkenbilt** [reply]
You are right -- I was mistaken when I said `/Info` goes under `/Root`. It goes in trailer as you have done. I'm glad you figured it out.

**Reply by SeeLook** [reply]
Anyway, You made my day. Once more thank You.
